File: elixir/rotational-cipher/lib/rotational_cipher.ex
defmodule RotationalCipher do
  @doc """
  Given a plaintext and amount to shift by, return a rotated string.

  Example:
  iex> RotationalCipher.rotate("Attack at dawn", 13)
  "Nggnpx ng qnja"
  """

  @spec rotate(text :: String.t(), shift :: integer) :: String.t()
  def rotate(text, shift) do
    text
    |> String.to_charlist()
    |> Enum.map(&translate(&1, shift))
    |> List.to_string()
  end

  defp translate(char, shift) when char in ?a..?z,
    do: ?a + rem((char - ?a) + shift, 26)

  defp translate(char, shift) when char in ?A..?Z,
    do: ?A + rem((char - ?A) + shift, 26)

  defp translate(char, _shift), do: char

end

File: kotlin/rotational-cipher/src/main/kotlin/RotationalCipher.kt
class RotationalCipher (val shift : Int) {

    fun encode(text: String): String {
        return text.map {
            when (it) {
                in 'a'..'z' -> 'a' + (it - 'a' + shift) % 26
                in 'A'..'Z' -> 'A' + (it - 'A' + shift) % 26
                else -> it
            
            }
        }.joinToString("")
    }
}
